src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -77,13 +77,17 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
                     return null;
                 }
 
+                var context = new CookieValidatePrincipalContext(Context, ticket, Options);
+
+                await Options.Notifications.ValidatePrincipal(context);
+
                 var allowRefresh = ticket.Properties.AllowRefresh ?? true;
                 if (issuedUtc != null && expiresUtc != null && Options.SlidingExpiration && allowRefresh)
                 {
                     var timeElapsed = currentUtc.Subtract(issuedUtc.Value);
                     var timeRemaining = expiresUtc.Value.Subtract(currentUtc);
 
-                    if (timeRemaining < timeElapsed)
+                    if (context.ShouldRenew || timeRemaining < timeElapsed)
```

Comment:
Should notification renewal be dependent on sliding refresh and all those conditions in the outer if?